aten/src/ATen/native/BinaryOps.cpp

```diff
@@ -613,6 +614,64 @@ Tensor& mul_(Tensor& self, const Scalar& other) {
   return at::mul_out(self, wrapped_scalar_tensor(other), self); // redispatch!
 }
 
+Device device_check(const Tensor& self, const Tensor& other) {
+  auto self_device = self.device();
+  auto other_device = other.device();
+  bool is_equal = (self_device == other_device);
+  if (!is_equal){
+    auto self_dim_is_zero = (self.dim() == 0);
+    auto other_dim_is_zero =  (other.dim() == 0);
+    if (self_dim_is_zero && other_dim_is_zero) {
+      if (self_device.is_cpu()) return other_device;
+    }
+    else if (self_dim_is_zero) {
+      TORCH_CHECK(self_device.is_cpu() && other_device.is_cuda());
+      return other_device;
+    }
+    if (other_dim_is_zero) {
+      TORCH_CHECK(other_device.is_cpu() && self_device.is_cuda());
+    }
+  }
+  return self_device;
+}
+
+Tensor mul_zerotensor(const Tensor& self, const Tensor& other) {
+  // hack to use the TensorIterator to get the correct broadcasting and type promotion logic
+  auto device_ = Device(DeviceType::Meta);
+  auto self_device = self.device();
+  auto meta_out = at::meta::mul(self.to(device_), other.to(device_));
+  if (self.is_meta() || other.is_meta()) {
```

Comment:
Is "zero tensors on the meta backend" a real use case?

I think there's probably not any reason to add special handling for the meta backend here, since `_efficientzerotensor` looks like it will do all of the right things anyway (and not be very expensive, since it also doesn't allocate data).

Also, totally optional but - if you want to eek out a bit of extra performance, you should be able to do something like:
```
auto meta_out = at::meta::mul(at::native::empty_strided_meta(self.sizes(), self.strides(), self.options()), at::native::empty_strided_meta(other.sizes(), other.strides(), other.options()));
```
to avoid redispatches in the `.to()` call.
```

Reply:
makes sense! removed it